File: server/node/conn.js
/**
 * @fileoverview WebSocket connection handle with token-bucket control-plane rate
 * limiting, native kernel backpressure monitoring, and zero-delay binary dispatching.
 */

import { WebSocket } from "ws";
import { create_message } from "./message.js";

const BACKPRESSURE = Object.freeze({
    DROP_NEWEST: 2,
    DROP_OLDEST: 1,
    DROP_SLOW_CLIENT: 0
});

/**
 * Creates an active WebSocket connection wrapper.
 *
 * @param {string} id - Connection UUID.
 * @param {WebSocket|object} ws - Active ws socket instance.
 * @param {object} hub - Owning hub coordinator.
 * @param {object} [claims] - Authenticated handshake claims.
 * @param {number} [capacity=8192] - Max queued frames buffer ceiling.
 * @param {number} [backpressure=0] - Backpressure strategy enum.
 * @param {number} [control_rate_limit=10] - Token-bucket refill rate for control events (tokens/sec).
 * @param {number} [control_burst=20] - Token-bucket max burst capacity for control events.
 * @returns {Readonly<object>} Frozen connection instance.
 */
function create_conn(id, ws, hub, claims, capacity, backpressure, control_rate_limit, control_burst) {
    const conn_claims = (
        typeof claims === "object" && claims !== null
        ? claims
        : Object.create(null)
    );

    // 32MB maximum buffer ceiling before slow-client eviction
    const max_buffer_bytes = (
        typeof capacity === "number" && capacity > 0
        ? capacity * 4096
        : 32 * 1024 * 1024
    );

    const strategy = (
        typeof backpressure === "number"
        ? backpressure
        : BACKPRESSURE.DROP_SLOW_CLIENT
    );

    const control_rate = (
        typeof control_rate_limit === "number" && control_rate_limit > 0
        ? control_rate_limit
        : 10.0
    );

    const control_burst_val = (
        typeof control_burst === "number" && control_burst > 0
        ? control_burst
        : 20
    );

    const rooms = Object.create(null);
    let is_closed = false;
    let close_sent = false;
    let is_alive = true;
    let control_tokens = control_burst_val;
    let last_control_check = Date.now();
    let self;

    // Disable Nagle's algorithm for sub-millisecond real-time frame delivery
    if (
        ws !== null &&
        typeof ws === "object" &&
        ws._socket !== undefined &&
        typeof ws._socket.setNoDelay === "function"
    ) {
        ws._socket.setNoDelay(true);
    }

    function track_room(room) {
        rooms[room] = true;
    }

    function untrack_room(room) {
        delete rooms[room];
    }

    function is_in_room(room) {
        return rooms[room] === true;
    }

    function joined_rooms() {
        return Object.keys(rooms);
    }

    /**
     * Evaluates token-bucket rate limiter for control-plane requests (join/leave).
     *
     * @returns {boolean} True if event is allowed, false if rate limited.
     */
    function allow_control_event() {
        const now = Date.now();
        const elapsed_sec = (now - last_control_check) / 1000;
        last_control_check = now;

        control_tokens += elapsed_sec * control_rate;
        if (control_tokens > control_burst_val) {
            control_tokens = control_burst_val;
        }

        if (control_tokens >= 1.0) {
            control_tokens -= 1.0;
            return true;
        }

        return false;
    }

    // Heartbeat pong tracking
    if (
        ws !== null &&
        typeof ws === "object" &&
        typeof ws.on === "function"
    ) {
        ws.on("pong", function () {
            is_alive = true;
        });
    }

    function cleanup() {
        if (is_closed === true) {
            return;
        }
        is_closed = true;
        hub.leave_all_rooms(self);
        hub.remove_conn(id);
        if (
            close_sent === false &&
            ws !== null &&
            typeof ws === "object" &&
            (ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING) &&
            typeof ws.close === "function"
        ) {
            close_sent = true;
            try {
                ws.close();
            } catch (ignore) {}
        }
    }

    function close_with(code, reason) {
        if (is_closed === true || close_sent === true) {
            return;
        }
        close_sent = true;
        is_closed = true;
        hub.leave_all_rooms(self);
        hub.remove_conn(id);
        if (
            ws !== null &&
            typeof ws === "object" &&
            (ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING) &&
            typeof ws.close === "function"
        ) {
            try {
                const status_code = (typeof code === "number" && code >= 1000 ? code : 1000);
                const reason_str = (typeof reason === "string" ? reason : "");
                ws.close(status_code, reason_str);
            } catch (ignore) {}
        }
    }

    function try_send(msg) {
        if (is_closed === true || ws === null || ws.readyState !== WebSocket.OPEN) {
            return false;
        }

        // Native libuv backpressure monitoring
        if (typeof ws.bufferedAmount === "number" && ws.bufferedAmount > max_buffer_bytes) {
            hub.metrics.onMessageDropped();

            if (strategy === BACKPRESSURE.DROP_NEWEST || strategy === BACKPRESSURE.DROP_OLDEST) {
                return false;
            }

            // DropSlowClient: send RFC 6455 1008 Policy Violation close frame
            close_with(1008, "Slow client buffer overflow");
            return false;
        }

        if (typeof ws.send === "function") {
            ws.send(msg, { binary: true });
        }

        hub.metrics.onMessageSent(msg.length);
        return true;
    }

    function send_to_room(room, event, payload) {
        const msg = create_message(room, event, "", id, payload);
        hub.broadcast_room(id, msg);
    }

    function send_to_client(dst_id, event, payload) {
        const msg = create_message("root", event, dst_id, id, payload);
        const target = hub.get_conn(dst_id);
        if (target !== undefined) {
            target.try_send(msg.encode());
        } else {
            hub.send_direct_to_cluster(msg);
        }
    }

    function check_heartbeat() {
        if (is_alive === false) {
            close_with(1000, "Heartbeat timeout");
            return;
        }
        is_alive = false;
        if (ws !== null && ws.readyState === WebSocket.OPEN && typeof ws.ping === "function") {
            ws.ping();
        }
    }

    self = Object.freeze({
        allow_control_event,
        check_heartbeat,
        claims: Object.freeze(conn_claims),
        cleanup,
        close_with,
        id,
        is_in_room,
        joined_rooms,
        send_to_client,
        send_to_room,
        track_room,
        try_send,
        untrack_room,
        ws
    });

    return self;
}

export {
    BACKPRESSURE,
    create_conn
};

File: server/go/conn.go
package roomer

import (
	"encoding/json"
	"net/http"
	"sync"
	"sync/atomic"
	"time"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
)

// Conn represents a single WebSocket connection with metadata, messaging channels, and rate limiting.
type Conn struct {
	ID            string
	Claims        map[string]string // Authenticated claims (e.g., user ID, roles)
	hub           *Hub
	send          chan []byte // Outbound message queue
	socket        *websocket.Conn
	cleanupOnce   sync.Once
	done          chan struct{}
	cleaningUp    int32
	roomsMu       sync.RWMutex
	rooms         map[string]struct{} // Set of joined room names
	config        Config
	controlMu     sync.Mutex
	controlTokens float64
	controlLast   time.Time
	writeMu       sync.Mutex
	closeMu       sync.Mutex
	closeSent     atomic.Bool
	closeCode     int
	closeReason   string
}

func (c *Conn) trackRoom(room string) bool {
	select {
	case <-c.done:
		return false
	default:
	}
	c.roomsMu.Lock()
	defer c.roomsMu.Unlock()
	select {
	case <-c.done:
		return false
	default:
		c.rooms[room] = struct{}{}
		return true
	}
}

func (c *Conn) untrackRoom(room string) {
	c.roomsMu.Lock()
	defer c.roomsMu.Unlock()
	delete(c.rooms, room)
}

func (c *Conn) joinedRooms() []string {
	c.roomsMu.RLock()
	defer c.roomsMu.RUnlock()
	rooms := make([]string, 0, len(c.rooms))
	for r := range c.rooms {
		rooms = append(rooms, r)
	}
	return rooms
}

// IsInRoom checks if this connection is currently tracked in a room.
func (c *Conn) IsInRoom(room string) bool {
	c.roomsMu.RLock()
	defer c.roomsMu.RUnlock()
	_, ok := c.rooms[room]
	return ok
}

// allowControlEvent evaluates the token-bucket rate limiter for control-plane requests (join/leave).
func (c *Conn) allowControlEvent() bool {
	c.controlMu.Lock()
	defer c.controlMu.Unlock()

	now := time.Now()
	elapsed := now.Sub(c.controlLast).Seconds()
	c.controlLast = now

	rate := c.config.ControlRateLimit
	if rate <= 0 {
		rate = 10.0 // default 10 operations per second
	}
	burst := float64(c.config.ControlBurst)
	if burst <= 0 {
		burst = 20.0 // default burst of 20
	}

	c.controlTokens += elapsed * rate
	if c.controlTokens > burst {
		c.controlTokens = burst
	}

	if c.controlTokens >= 1.0 {
		c.controlTokens -= 1.0
		return true
	}

	if c.config.Logger != nil {
		c.config.Logger.Warn("Control-plane rate limit exceeded for connection", "conn_id", c.ID)
	}
	return false
}

// TrySend attempts to send a binary message according to the configured BackpressureStrategy.
func (c *Conn) TrySend(msg []byte) bool {
	select {
	case <-c.done:
		return false
	default:
	}

	select {
	case <-c.done:
		return false
	case c.send <- msg:
		if c.config.Metrics != nil {
			c.config.Metrics.OnMessageSent(len(msg))
		}
		return true
	default:
		if c.config.Metrics != nil {
			c.config.Metrics.OnMessageDropped()
		}

		switch c.config.Backpressure {
		case DropOldest:
			select {
			case <-c.send: // Evict oldest queued message
			default:
			}
			select {
			case c.send <- msg:
				if c.config.Metrics != nil {
					c.config.Metrics.OnMessageSent(len(msg))
				}
				return true
			default:
				return false
			}

		case DropNewest:
			if c.config.Logger != nil {
				c.config.Logger.Warn("Conn dropped newest frame (buffer saturated)", "conn_id", c.ID)
			}
			return false

		case DropSlowClient:
			fallthrough
		default:
			if c.config.Logger != nil {
				c.config.Logger.Warn("Conn dropped message (buffer full or slow client)", "conn_id", c.ID)
			}
			// Trigger teardown with RFC 6455 policy violation close code
			if atomic.CompareAndSwapInt32(&c.cleaningUp, 0, 1) {
				go c.CloseWith(websocket.ClosePolicyViolation, "slow client buffer overflow")
			}
			return false
		}
	}
}

// SendToRoom broadcasts a message to all members of the specified room except self.
func (c *Conn) SendToRoom(roomName, event string, payload []byte) {
	msg := NewMessage(roomName, event, "", c.ID, payload)
	if room, ok := c.hub.getRoom(roomName); ok {
		room.emit(c, msg)
	}
}

// SendToClient sends a direct message to another client by connection ID using targeted unicast routing.
func (c *Conn) SendToClient(dstID, event string, payload []byte) {
	msg := NewMessage("root", event, dstID, c.ID, payload)
	if dst, ok := c.hub.getConn(dstID); ok {
		dst.TrySend(msg.Bytes())
	} else {
		c.hub.sendDirectToCluster(dstID, msg)
	}
}

// CloseWith sends a WebSocket close frame with an RFC 6455 status code and reason, then cleans up.
func (c *Conn) CloseWith(code int, reason string) {
	c.closeMu.Lock()
	if code == 0 {
		code = websocket.CloseNormalClosure
	}
	c.closeCode = code
	c.closeReason = reason
	c.closeMu.Unlock()

	if c.closeSent.CompareAndSwap(false, true) {
		_ = c.write(websocket.CloseMessage, websocket.FormatCloseMessage(code, reason))
	}
	c.cleanup()
}

// dispatch routes an incoming message to handlers, direct recipients, or rooms.
func (c *Conn) dispatch(msg *Message) {
	select {
	case <-c.done:
		return
	default:
	}

	msg.Src = c.ID
	msg.SrcLength = len(c.ID)

	switch msg.Event {
	case "join":
		if !c.allowControlEvent() {
			return
		}
		if c.config.RoomAuthorize != nil && !c.config.RoomAuthorize(c, msg.Room) {
			if c.config.Logger != nil {
				c.config.Logger.Warn("Unauthorized room join attempt", "conn_id", c.ID, "room", msg.Room)
			}
			return
		}
		c.hub.joinRoom(msg.Room, c)
		members := []byte("[]")
		snap := c.hub.getClusterPresence(msg.Room)
		if snapJSON, err := json.Marshal(snap); err == nil {
			members = snapJSON
		}
		ack := NewMessage(msg.Room, "join_ack", "", c.ID, members).Bytes()
		c.TrySend(ack)

	case "leave":
		if !c.allowControlEvent() {
			return
		}
		c.hub.leaveRoom(msg.Room, c)
		ack := NewMessage(msg.Room, "leave_ack", "", c.ID, []byte(c.ID)).Bytes()
		c.TrySend(ack)

	default:
		// Direct targeted messaging (local or cross-node unicast)
		if msg.Dst != "" {
			if dst, ok := c.hub.getConn(msg.Dst); ok {
				dst.TrySend(msg.Bytes())
			} else {
				c.hub.sendDirectToCluster(msg.Dst, msg)
			}
			return
		}

		// Custom handler dispatch
		if handler := getHandler(msg.Event); handler != nil {
			go func() {
				if err := handler(c, msg); err != nil {
					if c.config.Logger != nil {
						c.config.Logger.Error("Message handler execution error", "event", msg.Event, "conn_id", c.ID, "err", err)
					}
				}
			}()
			return
		}

		// Default room broadcast
		if room, ok := c.hub.getRoom(msg.Room); ok {
			room.emit(c, msg)
		}
	}
}

// cleanup safely removes the connection from all rooms, unregisters from cluster node registry, and releases resources.
func (c *Conn) cleanup() {
	c.cleanupOnce.Do(func() {
		close(c.done)
		c.hub.leaveAllRooms(c)
		c.hub.removeConn(c.ID)
		if c.socket != nil {
			_ = c.socket.Close()
		}
	})
}

// readPump reads messages from the WebSocket and dispatches them with early size enforcement.
func (c *Conn) readPump() {
	if c.socket == nil {
		return
	}
	defer c.cleanup()

	c.socket.SetReadLimit(c.config.MaxMessageSize)
	_ = c.socket.SetReadDeadline(time.Now().Add(c.config.PongWait))
	c.socket.SetPongHandler(func(string) error {
		_ = c.socket.SetReadDeadline(time.Now().Add(c.config.PongWait))
		return nil
	})

	for {
		_, data, err := c.socket.ReadMessage()
		if err != nil {
			return
		}
		if int64(len(data)) > c.config.MaxMessageSize {
			if c.config.Logger != nil {
				c.config.Logger.Warn("Message exceeded MaxMessageSize", "conn_id", c.ID, "size", len(data), "max", c.config.MaxMessageSize)
			}
			c.CloseWith(websocket.CloseMessageTooBig, "message too big")
			return
		}
		if c.config.Metrics != nil {
			c.config.Metrics.OnMessageReceived(len(data))
		}
		msg := BytesToMessageWithLimit(data, c.config.MaxMessageSize)
		if msg == nil {
			if c.config.Logger != nil {
				c.config.Logger.Warn("Malformed binary packet received", "conn_id", c.ID)
			}
			c.CloseWith(websocket.CloseProtocolError, "malformed binary packet")
			return
		}
		c.dispatch(msg)
	}
}

// write writes a message with a specified WebSocket message type and deadline under synchronization.
func (c *Conn) write(mt int, payload []byte) error {
	if c.socket == nil {
		return nil
	}
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	_ = c.socket.SetWriteDeadline(time.Now().Add(c.config.WriteWait))
	return c.socket.WriteMessage(mt, payload)
}

// writePump sends messages from the send channel and periodic pings.
func (c *Conn) writePump() {
	ticker := time.NewTicker(c.config.PingPeriod)
	defer func() {
		ticker.Stop()
		c.cleanup()
	}()

	for {
		select {
		case msg, ok := <-c.send:
			if !ok {
				if c.closeSent.CompareAndSwap(false, true) {
					c.closeMu.Lock()
					code := c.closeCode
					reason := c.closeReason
					c.closeMu.Unlock()
					if code == 0 {
						code = websocket.CloseGoingAway
						reason = "server shutting down"
					}
					_ = c.write(websocket.CloseMessage, websocket.FormatCloseMessage(code, reason))
				}
				return
			}
			if err := c.write(websocket.BinaryMessage, msg); err != nil {
				return
			}
		case <-c.done:
			if c.closeSent.CompareAndSwap(false, true) {
				c.closeMu.Lock()
				code := c.closeCode
				reason := c.closeReason
				c.closeMu.Unlock()
				if code == 0 {
					code = websocket.CloseGoingAway
					reason = "server shutting down"
				}
				_ = c.write(websocket.CloseMessage, websocket.FormatCloseMessage(code, reason))
			}
			return
		case <-ticker.C:
			if err := c.write(websocket.PingMessage, nil); err != nil {
				return
			}
		}
	}
}

// newConnection upgrades an HTTP request to a WebSocket and initializes a Conn.
func newConnection(w http.ResponseWriter, r *http.Request, claims map[string]string, cfg Config, h *Hub) *Conn {
	upgrader := websocket.Upgrader{
		ReadBufferSize:  cfg.ReadBufferSize,
		WriteBufferSize: cfg.WriteBufferSize,
		CheckOrigin:     cfg.CheckOrigin,
	}
	sock, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		return nil
	}
	id, err := uuid.NewRandom()
	if err != nil {
		_ = sock.Close()
		return nil
	}
	burst := float64(cfg.ControlBurst)
	if burst <= 0 {
		burst = 20.0
	}
	return &Conn{
		ID:            id.String(),
		Claims:        claims,
		hub:           h,
		socket:        sock,
		send:          make(chan []byte, cfg.ChannelCapacity),
		done:          make(chan struct{}),
		rooms:         make(map[string]struct{}),
		config:        cfg,
		controlTokens: burst,
		controlLast:   time.Now(),
	}
}
